Libraries/LibC/grp.cpp

```diff
@@ -26,54 +26,51 @@
 
 #include <AK/String.h>
 #include <AK/Vector.h>
+#include <errno_numbers.h>
 #include <grp.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <sys/mman.h>
 #include <unistd.h>
 
 extern "C" {
 
-#define GRDB_STR_MAX_LEN 256
+static FILE* s_stream = nullptr;
+static unsigned s_line_number = 0;
+static struct group s_group;
 
-struct group_with_strings : public group {
-    char name_buffer[GRDB_STR_MAX_LEN];
-    char passwd_buffer[GRDB_STR_MAX_LEN];
-    char* members[32];
-    char members_buffer[32][32];
-};
-
-static FILE* __grdb_stream = nullptr;
-static unsigned __grdb_line_number = 0;
-static struct group_with_strings* __grdb_entry = nullptr;
+static String s_name;
+static String s_passwd;
+static Vector<String> s_members;
+static Vector<const char*> s_members_ptrs;
 
 void setgrent()
 {
-    __grdb_line_number = 0;
-    if (__grdb_stream) {
-        rewind(__grdb_stream);
+    s_line_number = 0;
+    if (s_stream) {
+        rewind(s_stream);
     } else {
-        __grdb_stream = fopen("/etc/group", "r");
-        if (!__grdb_stream) {
+        s_stream = fopen("/etc/group", "r");
+        if (!s_stream) {
             perror("open /etc/group");
         }
-        assert(__grdb_stream);
-        __grdb_entry = (struct group_with_strings*)mmap_with_name(nullptr, getpagesize(), PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, 0, 0, "setgrent");
     }
 }
 
 void endgrent()
 {
-    __grdb_line_number = 0;
-    if (__grdb_stream) {
-        fclose(__grdb_stream);
-        __grdb_stream = nullptr;
-    }
-    if (__grdb_entry) {
-        munmap(__grdb_entry, getpagesize());
-        __grdb_entry = nullptr;
+    s_line_number = 0;
+    if (s_stream) {
+        fclose(s_stream);
+        s_stream = nullptr;
     }
+
+    s_name = {};
+    s_passwd = {};
+    s_members = {};
+    s_members_ptrs = {};
```

Comment:
Nit: First clear the pointers, then the data pointed-to. In this case it doesn't matter though, as nothing could possibly happen in the "meantime", except maybe a debugger getting confused.

Reply:
Yeah I'm this case I'm clearing out the whole thing, so it doesn't matter.

Reply:
Switched the two around